File: Heap/Python/IPO.py
# Problem link: https://leetcode.com/problems/ipo/description/
# Time: O(nlogn)
# Space: O(n)
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        maxHeap, idx = [], 0

        store = sorted([(c, p) for c, p in zip(capital, profits)])

        for i in range(k):

            while idx < len(store) and store[idx][0] <= w:
                heapq.heappush(maxHeap, -store[idx][1])
                idx += 1
            
            if not maxHeap:
                break
            
            w -= heapq.heappop(maxHeap)
        
        return w

File: Heap/Java/IPO.java
// Problem link: https://leetcode.com/problems/ipo/description/
// Time: O(nlogn)
// Space: O(n)
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);
        int idx = 0;

        int[][] store = new int[capital.length][2];
        for(int i = 0; i < capital.length; i++){
            store[i][0] = capital[i];
            store[i][1] = profits[i];
        }

        Arrays.sort(store, (a, b) -> a[0] - b[0]);

        for(int i = 0; i < k; i++){
            while(idx < store.length && store[idx][0] <= w){
                maxHeap.offer(store[idx][1]);
                idx++;
            }

            if(maxHeap.isEmpty()){
                break;
            }

            w += maxHeap.poll();
        }

        return w;
    }
}
